modyn/supervisor/internal/grpc_handler.py

```diff
@@ -434,3 +462,173 @@ def seed_selector(self, seed: int) -> None:
+            res: EvaluationStatusResponse = self.evaluator.get_evaluation_status(req)
+
+            if not res.valid:
+                logger.warning(f"Evaluation {current_evaluation_id} is invalid at server:\n{res}\n")
+                current_dataset_info.end_counter(True)
+                continue
+
+            if res.blocked:
+                blocked_in_a_row += 1
+                if blocked_in_a_row >= 3:
+                    logger.warning(
+                        f"Evaluator returned {blocked_in_a_row} blocked responses in a row, cannot update status."
+                    )
+                    blocked_in_a_row = 0
+                    current_dataset_info.end_counter(True)
+                else:
+                    working_queue.appendleft(current_evaluation_id)
+                continue
+            blocked_in_a_row = 0
+
+            if res.exception:
+                logger.warning(f"Exception at evaluator occurred:\n{res.exception}\n\n")
+                current_dataset_info.end_counter(True)
+                continue
+            if not res.is_running:
+                current_dataset_info.end_counter(False)
+                continue
+            if res.state_available:
+                assert res.samples_seen, f"Inconsistent server response:\n{res}"
+
+                current_dataset_info.progress_counter(res.samples_seen)
+            if res.is_running:
+                logger.warning("Evaluator is not blocked and is running, but no state is available.")
+
+            working_queue.append(current_evaluation_id)
+            sleep(1)
+
+        logger.info("Evaluation completed ✅")
+        self.status_bar.update(demo="Evaluation completed")
+
+    def store_evaluation_results(
+        self,
+        eval_directory: pathlib.Path,
+        pipeline_id: int,
+        trigger_id: int,
+        evaluations: dict[int, EvaluationStatusTracker],
+    ) -> None:
+        if not self.connected_to_evaluator:
+            raise ConnectionError("Tried to wait for evaluation to finish, but not there is no gRPC connection.")
+
+        # TODO(#281): store results in a framework-specific format
+        for evaluation_id in evaluations:
+            dataset_id = evaluations[evaluation_id].dataset_id
+            file_name = f"{pipeline_id}_{trigger_id}_{dataset_id}.eval"
+
+            req = EvaluationResultRequest(evaluation_id=evaluation_id)
+            res: EvaluationResultResponse = self.evaluator.get_evaluation_result(req)
+
+            if not res.valid:
+                logger.warning(f"Cannot get the evaluation result for evaluation {evaluation_id}")
+                continue
+
+            with open(eval_directory / file_name, "w+", encoding="utf-8") as file:
+                file.write(f"evaluation_id: {evaluation_id}\n")
+                file.write(f"dataset_size: {evaluations[evaluation_id].dataset_size}\n")
+                file.write("metrics:\n")
```

Comment:
could you maybe store it in some json file instead of this raw text format? maybe similar to the jsons created by the criteo evaluation script? i think that would be super useful (and needed to visualize the results)

Reply:
Yes, done.